### PR TITLE
Automatically setup virtualenv for development.

### PR DESCRIPTION
**Pull request by facundobatista**

### CODE REVIEW

**Comment by chipaca** [comment]
Two things:
* please use `python3 -m venv` instead of `virtualenv`, it's more likely to be there :-)
* instead of specifying `ops` and `test` by hand, exclude the `env` directory
thanks

**Comment by facundobatista** [comment]
> Two things:
> 
>     * please use `python3 -m venv` instead of `virtualenv`, it's more likely to be there :-)

Sure! Had to explicitly put setuptools and wheel in the requirement files...

 
>     * instead of specifying `ops` and `test` by hand, exclude the `env` directory
>       thanks

Great idea!

File: Makefile [hunk 1]
```diff
@@ -12,30 +12,37 @@
...
+	@x=$$(find . -name env -prune -o -name \*.py -not -empty -type f -print0 | xargs -0 grep -L "^# Copyright"); \
 	if [ "$$x" ]; then                                                                       \
 		echo "Please add copyright headers to the following files:";                     \
 		echo "$$x";                                                                      \
 		exit 1;                                                                          \
 	fi >&2
 
+$(ENV):
+	python3 -m venv $(ENV)
+	${PYTHON3} -m pip install -r requirements-test.txt
```
Comment (jameinel) [inline Makefile]: is that a common spelling? charms.reactive used "test_requirements.txt".
I don't have much stake in it, just don't want it to be surprising for people.

File: requirements-test.txt [hunk 2]
```diff
@@ -0,0 +1,4 @@
+-r requirements.txt
+autopep8
+flake8
+wheel  # needed by autopep8
```
Comment (jameinel) [inline requirements-test.txt]: I thought the point of requirements was to pull in the recursive dependencies. But if we need it, I suppose it is good to be explicit.
I'm a bit surprised that autopep8 depends on wheel.

File: Makefile (hunk 1 above)
Comment (facundobatista) [reply Makefile]: We just talked about this with John. The nice thing about start with "requirem.." is that both files are grouped to the eye when searching for them... Like, a developer searching if the project has requirement file(s), their will find both at once.

Comment (facundobatista) [reply Makefile]: That said, in other projects I used `requirements-dev.txt`

File: requirements-test.txt (hunk 2 above)
Comment (facundobatista) [reply requirements-test.txt]: I'm also surprised about that building part...

File: Makefile (hunk 1 above)
Comment (chipaca) [reply Makefile]: a search on github finds an equal number (~30k) of `test_requirements.txt`, `test-requirements.txt`, `requirements_test.txt`, and `requirements-test.txt`, FWIW.

File: requirements-test.txt (hunk 2 above)
Comment (chipaca) [reply requirements-test.txt]: I suspect the problem is because those projects don't declare these dependencies explicitly

**Comment by chipaca** [comment]
when I run `make` with this I still get angry errors about `bdist_wheel`,

```
Building wheels for collected packages: PyYAML, autopep8
  Running setup.py bdist_wheel for PyYAML ... error
  Complete output from command /home/john/canonical/operator/env/bin/python3 -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-4qekvir7/PyYAML/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /tmp/pip-wheel-y605gz0j --python-tag cp38:
  usage: -c [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
     or: -c --help [cmd1 cmd2 ...]
     or: -c --help-commands
     or: -c cmd --help
  
  error: invalid command 'bdist_wheel'
  
  ----------------------------------------
  Failed building wheel for PyYAML
  Running setup.py clean for PyYAML
  Running setup.py bdist_wheel for autopep8 ... error
  Complete output from command /home/john/canonical/operator/env/bin/python3 -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-4qekvir7/autopep8/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /tmp/pip-wheel-rqb582so --python-tag cp38:
  usage: -c [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
     or: -c --help [cmd1 cmd2 ...]
     or: -c --help-commands
     or: -c cmd --help
  
  error: invalid command 'bdist_wheel'
  
  ----------------------------------------
  Failed building wheel for autopep8
  Running setup.py clean for autopep8
Failed to build PyYAML autopep8
```

**Comment by chipaca** [comment]
Can I ask for one small change?

Make an `env: $(ENV)` target, and have an `install` step in `.travis.yml` that does `make env`.

This way the job failure mode when setting up the environment fails will be distinct from the failure mode when the tests themselves fail.

**Comment by facundobatista** [comment]
We're deferring how we'll handle dependencies in development until we understand the best way to handle dependencies in production, how charms will incorporate the operator framework and how everything will be "packaged". In the meantime I'll propose another branch improving the README to explicit the needed dependencies.

**Comment by jameinel** [comment]
Have we gotten any traction over taking over "ops" on Pypi ?

**Pull request closed without merging**